### ISSUE TITLE
l3keys: wishlist: code hook for variable setting action

### ISSUE THREAD

**Issue by callegar** [latex3/latex3#505]
I am missing the ability to have some code run *after* some variable is set.

As an example, suppose that I have a key associated to a variable setting action. It would be useful to me to have some code run right after the variable is set. Obviously, I can use `.code:n` but because `.code:n` cannot be used together with the variable setting action, if I use `.code:n`, I need to do the parsing of the value and its validation myself (e.g., suppose that the variable is boolean).  Otherwise, using the standard variable setting actions, after the `keys_set:nn` I need to check *all* the variables to see which ones have changed.

I think that it would be great to have a way to associate to a key that has a variable setting action (e.g. such as `bool_set:N`) a code hook without any parameter that is run *after* the variable setting. One possibility would be to have a `.code_append` property, that works like `.code`, but appends a code action rather than substituting the action that was previously defined.

Am I missing something?

**Comment by eg9** [comment]
That's interesting. Do you have some use case to show?

**Comment by Skillmon** [comment]
A possible implementation could be https://tex.stackexchange.com/a/466018/117050

**Comment by josephwright** [comment]
There is an obvious conceptual question here. Whilst we do have `.code:n`, most key settings are of the 'property' mindset: they are things one  can see as properties of a key, rather than actions carried out by a key. As such, one normally expects them to be queried at point-of-use, not do 'stuff' at point-of-setting.

Practically, there are various ways one might add a `.post_action:N` (presuming that a pre-action would also be needed). I'd rather avoid the `pgf` business of a potential change of stuff ....  My main concern there is I guess performance: I'd need to think about how best to avoid having to do a look-up at every step.

**Comment by josephwright** [comment]
As @eg9  says, I think a use case would be good here.

**Comment by wspr** [comment]
I agree with Joseph's (implicit) suggestion that it is better not to use keyval processing to execute package code. Logically separating the code into *gather options* then *execute options* results (in my experience) in code with a cleaner structure and fewer chances for faulty interactions between keys.

OTOH, I definitely take the point that some keys need to do more stuff than just set a variable, but if they also set a variable then it would be nice to indicate that somehow. As a completely random example from fontspec: (the first one I found)

```
\@@_keys_define_code:nnn {fontspec-preparse-external} {Extension}
  {
    \tl_set:Nn \l_@@_extension_tl {#1}
    \bool_if:NF \l_@@_external_bool
      {
        \keys_set:nn {fontspec-preparse-external} {Path}
      }
  }
```

But I suspect if I moved this test out of the keyval area and into the main package logic it would be a little more clear how and why this secondary code path was being executed.

**Comment by callegar** [comment]
Here is my use case. Beforehand, I'd like to make a disclaimer. I am just trying l3keys for the first times and I am not an expert in (x)keyval either. Hence, what I am trying to do may well be better done in other ways. In this case, corrections and suggestions on my approach are more than welcome!

1. Keys that are not of the "property" mindset, but more of the "action" mindset, for which `.code` would be OK but where I would like to take advantage of the validation and parsing that is already coded for boolean properties (and possibly also int, lenghts, etc.)

    In my mind, this could be handled not just by providing a code hook property for keys that already are associated to the property setting actions, but also in the opposite way, that is by being able to associate a validator property to keys that already have a `.code` action, so that the `#1` that `.code` gets is pre-validated and parsed into the correct data type, or by providing helper functions to be invoked in `.code` to do the validation/parsing of `#1` into the various data types.

2. Lots of keys (typically boolean or integers) that are used to modify the display of stuff encapsulated into ad-hoc semantic markup macros. For instance `colorset=1`, `newAddressSpacing=true`, `newLogoPosition=false`.

    When one of these is changed, it is easier to switch some global variables holding measurements, colors to different values (or even using `\let` to make some commands point to different actions) than having commands full of conditional paths. So you want to be able to perform an action (e.g switching many color names to new color values), but at the same time to have the key value parsed and validated and its name saved into a variable, because in some cases you still may need to use conditional paths based on it.

Hope this clarifies.

**Comment by josephwright** [comment]
I'm keen to see @FrankMittelbach's take here. To me, this looks like stuff for templates (or whatever we replace them with). My concern is that the order of keys becomes important if some of them reset stuff. Coding that up 'yourself' using `.code:n` is one thing (can be documented/predictable), but it's different if we provide an _ad hoc_ approach.

**Comment by callegar** [comment]
Possibly a `.post-setting-hook:n` would not be much different from `.code:n` for what concerns the order of operations and the possibility to document it.

In fact, I think that `.code:n` could be made to substitute any `.post-setting-hook:n` (whatever its actual name) by merely assuring the provided code does not *cancel* the variable setting property for the key but *stores* code to be executed either as the sole action or as an action to be taken *after* the variable setting if there is a variable setting property. In this way, if there is no variable setting property, then `.code:n` can work no differently than it does now and if there is a variable setting property for the key, then the code in `.code:n` can use the variable that has just been set.  This would avoid the need to introducing new key properties in the interface at all.
